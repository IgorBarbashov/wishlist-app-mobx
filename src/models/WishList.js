import { types, getParent, destroy } from 'mobx-state-tree';

// const data = {
//     "name": "Toy-1",
//     "price": 30.53,
//     "image": "https://lh3.googleusercontent.com/TFuORPSuxYy8PQKFV2UBNHDr-aKclxqWIlLgx3F7ORId4KU6u_CTBpcY-hZ5e_B2NiPHv3MJ7PtLK6FdARZwu1oPuYHx9gqLrDu5rt4"
// };

export const WishListItem = types.model({
    name: types.string, // it will be @observable property
    price: types.number, // it will be @observable property
    image: "" // it will be @observable property
}).actions(self => ({
    changeName(newName) { // it will be @action
        self.name = newName;
    },
    changePrice(newPrice) { // it will be @action
        self.price = newPrice;
    },
    changeImage(newImage) { // it will be @action
        self.image = newImage;
    },
    remove() {
        getParent(self, 2).remove(self)
    }
}));

export const WishList = types.model({
    items: types.optional(types.array(WishListItem), [])
}).actions(self => ({
    addItem(item) {
        self.items.push(item);
    },
    remove(item) {
        destroy(item);
        // self.items.splice(self.items.indexOf(item), 1);
    }
})).views(self => ({
    get totalPrice() { // it will be @computed property
        return self.items.reduce((acc, el) => acc + el.price, 0);
    }
}));