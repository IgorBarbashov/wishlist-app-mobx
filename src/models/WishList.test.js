import { getSnapshot, onSnapshot, onPatch } from 'mobx-state-tree';
import { WishListItem, WishList } from './WishList';
import { reaction } from 'mobx';

it("can create a instance of a model", () => {
    const item = WishListItem.create({
        "name": "Toy-1",
        "price": 30.53,
        "image": "https://lh3.googleusercontent.com/TFuORPSuxYy8PQKFV2UBNHDr-aKclxqWIlLgx3F7ORId4KU6u_CTBpcY-hZ5e_B2NiPHv3MJ7PtLK6FdARZwu1oPuYHx9gqLrDu5rt4"
    });

    expect(item.price).toBe(30.53);
    item.changeName('Toyi-3');
    expect(item.name).toBe('Toyi-3');
});

it("can create a wishlist", () => {
    const list = WishList.create({
        items: [
            {
                "name": "Toy-1",
                "price": 30.53,
                "image": "https://lh3.googleusercontent.com/TFuORPSuxYy8PQKFV2UBNHDr-aKclxqWIlLgx3F7ORId4KU6u_CTBpcY-hZ5e_B2NiPHv3MJ7PtLK6FdARZwu1oPuYHx9gqLrDu5rt4"
            }
        ]
    });

    expect(list.items.length).toBe(1);
    expect(list.items[0].price).toBe(30.53);
});

it("can add a new items", () => {
    const list = WishList.create();
    expect(list.items.length).toBe(0);

    const states = [];
    onSnapshot(list, snapshot => {
        states.push(snapshot);
    });

    list.addItem({
        "name": "Toy-1",
        "price": 30.53,
        "image": "https://lh3.googleusercontent.com/TFuORPSuxYy8PQKFV2UBNHDr-aKclxqWIlLgx3F7ORId4KU6u_CTBpcY-hZ5e_B2NiPHv3MJ7PtLK6FdARZwu1oPuYHx9gqLrDu5rt4"
    });

    expect(list.items.length).toBe(1);
    expect(list.items[0].price).toBe(30.53);
    list.items[0].changeName('Toy-100');
    expect(list.items[0].name).toBe('Toy-100');

    expect(getSnapshot(list)).toMatchSnapshot();

    expect(states).toMatchSnapshot();
});

it("can add a new items - 2", () => {
    const list = WishList.create();
    expect(list.items.length).toBe(0);

    const patches = [];
    onPatch(list, patch => {
        patches.push(patch);
    });

    list.addItem({
        "name": "Toy-1",
        "price": 30.53,
        "image": "https://lh3.googleusercontent.com/TFuORPSuxYy8PQKFV2UBNHDr-aKclxqWIlLgx3F7ORId4KU6u_CTBpcY-hZ5e_B2NiPHv3MJ7PtLK6FdARZwu1oPuYHx9gqLrDu5rt4"
    });

    expect(list.items.length).toBe(1);
    expect(list.items[0].price).toBe(30.53);
    list.items[0].changeName('Toy-100');
    expect(list.items[0].name).toBe('Toy-100');

    expect(getSnapshot(list)).toMatchSnapshot();

    expect(patches).toMatchSnapshot();
});

it("can calculate totalPrice", () => {
    const list = WishList.create({
        items: [
            {
                "name": "Toy-1",
                "price": 30.53,
                "image": "https://lh3.googleusercontent.com/TFuORPSuxYy8PQKFV2UBNHDr-aKclxqWIlLgx3F7ORId4KU6u_CTBpcY-hZ5e_B2NiPHv3MJ7PtLK6FdARZwu1oPuYHx9gqLrDu5rt4"
            }
        ]
    });

    list.addItem({
        "name": "Toy-1",
        "price": 30.53,
        "image": "https://lh3.googleusercontent.com/TFuORPSuxYy8PQKFV2UBNHDr-aKclxqWIlLgx3F7ORId4KU6u_CTBpcY-hZ5e_B2NiPHv3MJ7PtLK6FdARZwu1oPuYHx9gqLrDu5rt4"
    });

    expect(list.items.length).toBe(2);
    expect(list.items[0].price).toBe(30.53);
    list.items[0].changeName('Toy-100');
    expect(list.items[0].name).toBe('Toy-100');
    expect(list.totalPrice).toBe(61.06);

    let changed = 0;
    reaction(() => list.totalPrice, () => changed++ );
    
    expect(changed).toBe(0);
    console.log(list.totalPrice);
    list.items[0].changeName('ToyToy');
    expect(changed).toBe(0);
    list.items[0].changePrice(10);
    expect(changed).toBe(1);
});