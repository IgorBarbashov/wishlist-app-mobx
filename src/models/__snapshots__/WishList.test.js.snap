// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add a new items - 2 1`] = `
Object {
  "items": Array [
    Object {
      "image": "https://lh3.googleusercontent.com/TFuORPSuxYy8PQKFV2UBNHDr-aKclxqWIlLgx3F7ORId4KU6u_CTBpcY-hZ5e_B2NiPHv3MJ7PtLK6FdARZwu1oPuYHx9gqLrDu5rt4",
      "name": "Toy-100",
      "price": 30.53,
    },
  ],
}
`;

exports[`can add a new items - 2 2`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "https://lh3.googleusercontent.com/TFuORPSuxYy8PQKFV2UBNHDr-aKclxqWIlLgx3F7ORId4KU6u_CTBpcY-hZ5e_B2NiPHv3MJ7PtLK6FdARZwu1oPuYHx9gqLrDu5rt4",
      "name": "Toy-1",
      "price": 30.53,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "Toy-100",
  },
]
`;

exports[`can add a new items 1`] = `
Object {
  "items": Array [
    Object {
      "image": "https://lh3.googleusercontent.com/TFuORPSuxYy8PQKFV2UBNHDr-aKclxqWIlLgx3F7ORId4KU6u_CTBpcY-hZ5e_B2NiPHv3MJ7PtLK6FdARZwu1oPuYHx9gqLrDu5rt4",
      "name": "Toy-100",
      "price": 30.53,
    },
  ],
}
`;

exports[`can add a new items 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "https://lh3.googleusercontent.com/TFuORPSuxYy8PQKFV2UBNHDr-aKclxqWIlLgx3F7ORId4KU6u_CTBpcY-hZ5e_B2NiPHv3MJ7PtLK6FdARZwu1oPuYHx9gqLrDu5rt4",
        "name": "Toy-1",
        "price": 30.53,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "https://lh3.googleusercontent.com/TFuORPSuxYy8PQKFV2UBNHDr-aKclxqWIlLgx3F7ORId4KU6u_CTBpcY-hZ5e_B2NiPHv3MJ7PtLK6FdARZwu1oPuYHx9gqLrDu5rt4",
        "name": "Toy-100",
        "price": 30.53,
      },
    ],
  },
]
`;

exports[`can calculate totalPrice 1`] = `
Object {
  "items": Array [
    Object {
      "image": "https://lh3.googleusercontent.com/TFuORPSuxYy8PQKFV2UBNHDr-aKclxqWIlLgx3F7ORId4KU6u_CTBpcY-hZ5e_B2NiPHv3MJ7PtLK6FdARZwu1oPuYHx9gqLrDu5rt4",
      "name": "Toy-100",
      "price": 30.53,
    },
    Object {
      "image": "https://lh3.googleusercontent.com/TFuORPSuxYy8PQKFV2UBNHDr-aKclxqWIlLgx3F7ORId4KU6u_CTBpcY-hZ5e_B2NiPHv3MJ7PtLK6FdARZwu1oPuYHx9gqLrDu5rt4",
      "name": "Toy-1",
      "price": 30.53,
    },
  ],
}
`;
